rust-runtime/aws-smithy-runtime-api/src/client/http/request.rs

```diff
@@ -0,0 +1,476 @@
+pub struct Uri {
+    as_string: String,
+    parsed: http0::Uri,
+}
+
+impl TryFrom<String> for Uri {
+    type Error = HttpError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        let parsed = value.parse()?;
+        Ok(Uri {
+            as_string: value,
+            parsed,
+        })
+    }
+}
+
+impl From<http0::Uri> for Uri {
+    fn from(value: http::Uri) -> Self {
+        Self {
+            as_string: value.to_string(),
+            parsed: value,
+        }
+    }
+}
+
+impl<B> TryInto<http0::Request<B>> for Request<B> {
+    type Error = HttpError;
+
+    fn try_into(self) -> Result<http::Request<B>, Self::Error> {
+        Ok(self.into_http03x()?)
+    }
+}
+
+impl<B> Request<B> {
+    /// Converts this request into an http 0.x request.
+    ///
+    /// Depending on the internal storage type, this operation may be free or it may have an internal
+    /// cost.
+    pub fn into_http03x(self) -> Result<http0::Request<B>, HttpError> {
+        let mut req = http::Request::builder()
+            .uri(self.uri.parsed)
+            .method(self.method)
+            .body(self.body)
+            .expect("known valid");
+        let mut headers = HeaderMap::new();
+        headers.extend(
+            self.headers
+                .headers
+                .into_iter()
+                .map(|(k, v)| (k, v.into_http0())),
+        );
+        *req.headers_mut() = headers;
+        *req.extensions_mut() = self.extensions;
+        Ok(req)
+    }
+
+    /// Returns a GET request with no URI
+    pub fn new(body: B) -> Self {
+        Self {
+            body,
+            uri: Uri::from(http0::Uri::from_static("/")),
+            method: Method::GET,
+            extensions: Default::default(),
+            headers: Default::default(),
+        }
+    }
+
+    /// Returns a reference to the header map
+    pub fn headers(&self) -> &Headers {
+        &self.headers
+    }
+
+    /// Returns a mutable reference to the header map
+    pub fn headers_mut(&mut self) -> &mut Headers {
+        &mut self.headers
+    }
+
+    /// Returns the body associated with the request
+    pub fn body(&self) -> &B {
+        &self.body
+    }
+
+    /// Returns a mutable reference to the body
+    pub fn body_mut(&mut self) -> &mut B {
+        &mut self.body
+    }
+
+    /// Returns the method associated with this request
+    pub fn method(&self) -> &str {
+        &self.method.as_str()
+    }
+
+    /// Returns the URI associated with this request
+    pub fn uri(&self) -> &str {
+        &self.uri.as_string
+    }
+
+    /// Sets the URI of this request
+    pub fn set_uri<U>(&mut self, uri: U) -> Result<(), U::Error>
+    where
+        U: TryInto<Uri>,
+    {
+        let uri = uri.try_into().map_err(|e| e.into())?;
+        self.uri = uri;
+        Ok(())
+    }
+
+    /// Adds an extension to the request extensions
+    pub fn add_extension<T: Send + Sync + Clone + 'static>(&mut self, extension: T) {
```

Comment:
Should this be fallible in case the underlying HTTP library doesn't support extensions?

Reply:
so _currently_ we have fallibility in the conversion to the final request type—you actually don't know what the final request type is going to be until the request is dispatched.

We can probably refactor at some point to remove the need to set request extensions and instead handle it during the conversion to `http03x::Request`...not a bad idea actually, hmmm